azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,403 @@ def _run_op(self, op):
+                                    )
+                                )
+                        else:
+                            error = exceptions.ServiceError(
+                                "Query Status Operation encountered an invalid registration status {status} with a status code of {status_code}".format(
+                                    status=registration_status, status_code=success_status_code
+                                )
+                            )
+
+                        query_status_op.complete(error=error)
+
+                else:
+                    query_status_op.complete(error=error)
+
+            self.send_op_down(
+                pipeline_ops_base.RequestAndResponseOperation(
+                    request_type=constant.QUERY,
+                    method="GET",
+                    resource_location="/",
+                    query_params={"operation_id": query_status_op.operation_id},
+                    request_body=query_status_op.request_payload,
+                    callback=on_query_response,
+                )
+            )
+
+        else:
+            super(PollingStatusStage, self)._run_op(op)
+
+
+class RegistrationStage(CommonProvisioningStage):
+    """
+    This is the first stage that decides converts a registration request
+    into a normal request and response operation.
+    Upon the receipt of the response this stage decides whether
+    to send another registration request or send a query request.
+    Depending on the status and result of the response
+    this stage may also complete the registration process.
+    """
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, op):
+        if isinstance(op, pipeline_ops_provisioning.SendRegistrationRequestOperation):
+            initial_register_op = op
+
+            def on_registration_response(op, error):
+                logger.debug(
+                    "{stage_name}({op_name}): Received response with status code {status_code} for SendRegistrationRequestOperation".format(
+                        stage_name=self.name, op_name=op.name, status_code=op.status_code
+                    )
+                )
+
+                # This could be an error that has been reported to this stage Or this
+                # could be an error because the service responded with status code >=300
+                error = CommonProvisioningStage._get_error(op, error=error)
+
+                if not error:
+                    success_status_code = op.status_code
+                    retry_interval = (
+                        int(op.retry_after, 10)
+                        if op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
+                    decoded_response = CommonProvisioningStage._decode_response(op)
+                    operation_id = CommonProvisioningStage._get_operation_id(decoded_response)
```

Comment:
You can still access static methods with self

Reply:
what is the accepted pattern ? using `self._get_operation_id` or `CommonProvisioningStage._get_operation_id`

---
In reply to: [366024926](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r366024926) [](ancestors = 366024926)